### PR TITLE
Add Authz queries to QueryClient

### LINKED ISSUE #1080: Add Authz queries to QueryClient
We need to add `/cosmos/authz/v1beta1/grants` to the authz extension.

### PR DESCRIPTION
**Pull request by msteiner96**
Resolves #1080

### CODE REVIEW

File: packages/stargate/src/modules/authz/queries.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+import { createPagination, createProtobufRpcClient, QueryClient } from "../../queryclient";
+
+export interface AuthzExtension {
+  readonly authz: {
+    readonly grants: (
+      granter: string,
+      grantee: string,
+      msgTypeUrl: string,
+      pagination?: Uint8Array,
```
Comment (webmaster128) [inline packages/stargate/src/modules/authz/queries.ts:11]:
```suggestion
      paginationKey?: Uint8Array,
```

File: packages/stargate/src/modules/authz/queries.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+export function setupAuthzExtension(base: QueryClient): AuthzExtension {
+  // Use this service to get easy typed access to query methods
+  // This cannot be used for proof verification
+  const rpc = createProtobufRpcClient(base);
+  const queryService = new QueryClientImpl(rpc);
+
+  return {
+    authz: {
+      grants: async (granter: string, grantee: string, msgTypeUrl: string, pagination?: Uint8Array) => {
```
Comment (webmaster128) [inline packages/stargate/src/modules/authz/queries.ts:24]:
```suggestion
      grants: async (granter: string, grantee: string, msgTypeUrl: string, paginationKey?: Uint8Array) => {
```

File: packages/stargate/src/modules/authz/queries.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        const response = await queryService.Grants({
+          granter: granter,
+          grantee: grantee,
+          msgTypeUrl: msgTypeUrl,
+          pagination: createPagination(pagination),
```
Comment (webmaster128) [inline packages/stargate/src/modules/authz/queries.ts:29]:
```suggestion
          pagination: createPagination(paginationKey),
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -45,6 +45,7 @@ and this project adheres to
   wordlist is shipped. ([#966])
 - @cosmjs/cli: Rename binary `cosmwasm-cli` to `cosmjs-cli` ([#1033]).
 - @cosmjs/stargate & @cosmjs/cosmwasm-stargate: Removed default types from AminoTypes. ([1079])
+- @cosmjs/stargate: Added Authz Queries. ([#1080]).
```
Comment (webmaster128) [inline CHANGELOG.md:48]:
```suggestion
- @cosmjs/stargate: Added Authz queries. ([#1080]).
```

**Review by webmaster128 — COMMENTED** [review]
Nice!

Can we add a test for that? This won't work in simapp 42 but should be available in simapp 44.

**Comment by msteiner96** [comment]
@webmaster128 Regarding the tests: Is there already an example test case that I could query?

**Comment by webmaster128** [comment]
> @webmaster128 Regarding the tests: Is there already an example test case that I could query?

In `packages/stargate/src/modules/gov/queries.spec.ts` you see an example test where the test itself generates example data that can then be queried. We should have everything available to do something similar here and send a `MsgGrant` that creates a grant which can be found in the test.

File: packages/stargate/src/modules/authz/queries.spec.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        defaultFee,
+        "Test grant for simd",
+      );
+      assertIsDeliverTxSuccess(grantResult);
+      await sleep(75); // wait until transactions are indexed
+
+      client.disconnect();
+    }
+  });
+  describe("grants", () => {
```
Comment (webmaster128) [inline packages/stargate/src/modules/authz/queries.spec.ts:76]:
```suggestion

  describe("grants", () => {
```

File: packages/stargate/src/modules/authz/queries.spec.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  return [QueryClient.withExtensions(tmClient, setupAuthzExtension), tmClient];
+}
+
+describe("AuthzExtension", () => {
+  const defaultFee = {
+    amount: coins(25000, "ucosm"),
+    gas: "1500000", // 1.5 million
+  };
+  const granter1Address = faucet.address1;
+  const grantee1Address = faucet.address2;
```
Comment (webmaster128) [inline packages/stargate/src/modules/authz/queries.spec.ts:32]: Can we use a random address for this one which does not exist (see `makeRandomAddress` from `packages/cosmwasm-stargate/src/testutils.spec.ts`)? Then we get a new (granter, grantee) pair each time we run the test and exactly one expected result element.

File: packages/stargate/src/modules/authz/queries.spec.ts [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    }
+  });
+  describe("grants", () => {
+    it("works", async () => {
+      pendingWithoutSimapp44();
+      const [client, tmClient] = await makeClientWithAuthz(simapp.tendermintUrl);
+
+      const response = await client.authz.grants(granter1Address, grantee1Address, "");
+      expect(response).not.toBeUndefined();
+      expect(response).not.toBeNull();
```
Comment (webmaster128) [inline packages/stargate/src/modules/authz/queries.spec.ts:83]: Can you test the value of `response` here?

File: packages/stargate/src/modules/authz/queries.spec.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+  describe("grants", () => {
+    it("works", async () => {
+      pendingWithoutSimapp44();
+      const [client, tmClient] = await makeClientWithAuthz(simapp.tendermintUrl);
+      const response = await client.authz.grants(granter1Address, grantee1Address, "");
+      const grant = response.grants[0];
+
+      // Needs to respond with a grant
+      expect(grant.authorization).not.toBeUndefined();
+      expect(grant.authorization).not.toBeNull();
```
Comment (webmaster128) [inline packages/stargate/src/modules/authz/queries.spec.ts:89]: You can use `assertDefined(grant.authorization)` here to get the runtime check and type inference ath the same time.

File: packages/stargate/src/modules/authz/queries.spec.ts [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+      // Needs to be GenericAuthorization to decode it below
+      expect(grant.authorization?.typeUrl).toBe("/cosmos.authz.v1beta1.GenericAuthorization");
+
+      // If above is true
+      if (grant.authorization) {
```
Comment (webmaster128) [inline packages/stargate/src/modules/authz/queries.spec.ts:95]: then you don't need this

File: packages/stargate/src/modules/authz/queries.spec.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+      const [client, tmClient] = await makeClientWithAuthz(simapp.tendermintUrl);
+      const response = await client.authz.grants(granter1Address, grantee1Address, "");
+      const grant = response.grants[0];
+
+      // Needs to respond with a grant
+      expect(grant.authorization).not.toBeUndefined();
+      expect(grant.authorization).not.toBeNull();
+
+      // Needs to be GenericAuthorization to decode it below
+      expect(grant.authorization?.typeUrl).toBe("/cosmos.authz.v1beta1.GenericAuthorization");
```
Comment (webmaster128) [inline packages/stargate/src/modules/authz/queries.spec.ts:92]: Let's use `toEqual` consistently, even if it is no difference for strings

File: packages/stargate/src/modules/authz/queries.spec.ts [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+      // If above is true
+      if (grant.authorization) {
+        // Decode the message
+        const msgDecoded = GenericAuthorization.decode(grant.authorization.value).msg;
+
+        // Check if its the same one then we granted
+        expect(msgDecoded).toEqual(grantedMsg);
```
Comment (webmaster128) [inline packages/stargate/src/modules/authz/queries.spec.ts:100]: 🎉

**Review by webmaster128 — COMMENTED** [review]
Something is strange with the merging. We have the diff from other PRs in here. If you want you can just ignore and I fix it when merging.

File: packages/stargate/src/modules/authz/queries.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+import { createPagination, createProtobufRpcClient, QueryClient } from "../../queryclient";
+
+export interface AuthzExtension {
+  readonly authz: {
+    readonly grants: (
+      granter: string,
+      grantee: string,
+      msgTypeUrl: string,
+      paginationLey?: Uint8Array,
```
Comment (webmaster128) [inline packages/stargate/src/modules/authz/queries.ts:11]:
```suggestion
      paginationKey?: Uint8Array,
```

File: packages/stargate/src/modules/authz/queries.spec.ts [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+      client.disconnect();
+    }
+  });
+
+  describe("grants", () => {
+    it("works", async () => {
+      pendingWithoutSimapp44();
+      const [client, tmClient] = await makeClientWithAuthz(simapp.tendermintUrl);
+      const response = await client.authz.grants(granter1Address, grantee1Address, "");
+      const grant = response.grants[0];
```
Comment (webmaster128) [inline packages/stargate/src/modules/authz/queries.spec.ts:85]:
```suggestion
      expect(response.grants.length).toEqual(1);
      const grant = response.grants[0];
```

File: packages/stargate/src/modules/authz/queries.spec.ts [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+
+      // Needs to respond with a grant
+      assertDefined(grant.authorization);
+
+      // Needs to be GenericAuthorization to decode it below
+      expect(grant.authorization?.typeUrl).toEqual("/cosmos.authz.v1beta1.GenericAuthorization");
```
Comment (webmaster128) [inline packages/stargate/src/modules/authz/queries.spec.ts:91]:
```suggestion
      expect(grant.authorization.typeUrl).toEqual("/cosmos.authz.v1beta1.GenericAuthorization");
```

**Review by webmaster128 — APPROVED** [review]

**Pull request merged**